src/device_handle/async_api.rs

```diff
@@ -0,0 +1,266 @@
+use crate::{DeviceHandle, UsbContext};
+use libusb1_sys as ffi;
+
+use libc::c_void;
+use std::collections::VecDeque;
+use std::convert::{TryFrom, TryInto};
+use std::marker::PhantomPinned;
+use std::pin::Pin;
+use std::ptr::NonNull;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+use thiserror::Error;
+
+type CompletedQueue = Arc<Mutex<VecDeque<usize>>>;
+type Transfer<C> = Pin<Box<AsyncTransfer<C>>>;
+
+#[derive(Error, Debug)]
+pub enum AsyncError {
+    #[error("Transfer timed out")]
+    TransferTimeout,
+    #[error("Poll timed out")]
+    PollTimeout,
+    #[error("Transfer is stalled")]
+    Stall,
+    #[error("Device was disconnected")]
+    Disconnected,
+    #[error("Other Error: {0}")]
+    Other(&'static str),
+    #[error("{0}ERRNO: {1}")]
+    Errno(&'static str, i32),
+    #[error("Transfer was cancelled")]
+    Cancelled,
+}
+
+struct AsyncTransfer<C: UsbContext> {
+    ptr: NonNull<ffi::libusb_transfer>,
+    /// The ID of the transfer, as an idx into the transfer pool
+    pool_id: usize,
+    buffer: Vec<u8>,
+    completed_queue: CompletedQueue,
+    device: Arc<DeviceHandle<C>>,
+    // `ptr` holds a pointer to `buffer`, `device`, and `self`, so we must be `!Unpin`
+    _pin: PhantomPinned,
+}
+impl<C: UsbContext> AsyncTransfer<C> {
+    fn new_bulk(
+        pool_id: usize,
+        completed_queue: CompletedQueue,
+        device: Arc<DeviceHandle<C>>,
+        endpoint: u8,
+        buffer: Vec<u8>,
+        timeout: std::time::Duration,
+    ) -> Pin<Box<Self>> {
+        // non-isochronous endpoints (e.g. control, bulk, interrupt) specify a value of 0
+        // This is step 1 of async API
+        let ptr = unsafe { ffi::libusb_alloc_transfer(0) };
+        let ptr = NonNull::new(ptr).expect("Could not allocate transfer!");
+        let timeout = libc::c_uint::try_from(timeout.as_millis())
+            .expect("Duration was too long to fit into a c_uint");
```

Comment:
wasn't sure what a good default would be for this otherwise. One thing we could do is just have the poll timeout, and read timeouts would automatically be resubmitted without propagating the error to the user. That would simplify the API I think. 

Any ideas for reasonable defaults for that value?

Reply:
Doesn't setting the timeout to `0` mean "no timeout"?

Reply:
Yep, [youre right](https://libusb.sourceforge.io/api-1.0/structlibusb__transfer.html#a9a12af15ca5b482f5dcaebd26a848cbb). From the documentation of the `timeout` field:
> Timeout for this transfer in milliseconds.
>
> A value of 0 indicates no timeout.